app/src/main/java/unipd/se18/ocrcamera/AllergensManager.java

```diff
@@ -0,0 +1,254 @@
+     * @modify allergensList by putting all allergens in it
+     * @author Cervo Nicolò (g3)
+     */
+    public ArrayList<Allergen> getAllergensList() {
+
+        return allergensList;
+    }
+
+    /**
+     * parses through selected_allergens_sharedPref to return the list of the selected allergens
+     * @return selectedAllergensList, ArrayList<Allergen> of selected allergens
+     * @author Nicolò Cervo (g3)
+     */
+    public ArrayList<Allergen> getSelectedAllergensList() {
+
+        // Get the content of selected_allergens_sharedPref
+        Map<String, ?> selectedAllergens = sp.getAll();
+        ArrayList<Allergen> selectedAllergensList = new ArrayList<>();
+
+        // For every line in selected_allergens_sharedPref create the corresponding Allergen
+        // and store it in selectedAllergensList
+        for(Map.Entry<String, ?> entry : selectedAllergens.entrySet()) {
+
+            Allergen allergen = new Allergen();
+            allergen.setCommonName(entry.getKey());
+            allergen.setInciNames(entry.getValue().toString().split(", "));
+            selectedAllergensList.add(allergen);
+
+        }
+        return  selectedAllergensList;
+    }
+
+    /**
+     * checks if ingr is or derives from any allergen.
+     * @param ingr, ingredient (recognized form the Ocr text)
+     * @return foundAllergens, ArrayList<Allergen> containing the allergens found in the inciName or
+     *         in the description of ingr, empty list if none is found
+     * @author Cervo Nicolò
+     */
+    public ArrayList<Allergen> checkForAllergens(Ingredient ingr) {
+        ArrayList<Allergen> foundAllergens = new ArrayList<>();
+
+        for(Allergen allergen : allergensList) {
+
+            // for every inciName associated with this allergen
+            for(String allergenInciName : allergen.getInciNames()) {
+
+                // if the allergen commonName or inciName appears in the ingredient inciName or
+                // description respectively
+                if(ingr.getInciName().contains(allergenInciName) ||
+                        ingr.getDescription().toUpperCase().contains(allergen.getCommonName())) {
+
+                    // add the allergen to foundAllergens
+                    foundAllergens.add(allergen);
+
+                    // if the allergen is found go to the next one
+                    break;
+                }
+            }
+        }
+        return foundAllergens;
+    }
+
+    /**
+     * Check if an ingredient contains a selected allergen common name or inci name (stored in SharedPreferences sp)
+     * @param ingr, ingredient to check for allergens
+     * @return the allergen found in the ingredient, empty ArrayList if no allergen is found
+     * @author Nicolò Cervo (g3)
+     */
+    public ArrayList<Allergen> checkForSelectedAllergens(Ingredient ingr) {
+
+        Map<String, ?> selectedAllergens = sp.getAll();
+        Allergen allergen = new Allergen();
```

Comment:
I would suggest to declare this variable "allergen" into the "if" block inside the second "for" cycle, in order to allocate its memory only if it's actually needed,